sklearn/utils/extmath.py

```diff
@@ -255,37 +257,70 @@ def randomized_range_finder(
     analysis
     A. Szlam et al. 2014
     """
+    xp, is_array_api_compliant = get_namespace(A)
     random_state = check_random_state(random_state)
 
     # Generating normal random vectors with shape: (A.shape[1], size)
-    Q = random_state.normal(size=(A.shape[1], size))
-    if hasattr(A, "dtype") and A.dtype.kind == "f":
-        # Ensure f32 is preserved as f32
-        Q = Q.astype(A.dtype, copy=False)
+    # XXX: generate random number directly from xp if it's possible
+    # one day.
+    Q = xp.asarray(random_state.normal(size=(A.shape[1], size)))
+    if hasattr(A, "dtype") and xp.isdtype(A.dtype, kind="real floating"):
+        # Use float32 computation and components if A has a float32 dtype.
+        Q = xp.astype(Q, A.dtype, copy=False)
+
+    # Move Q to device if needed only after converting to float32 if needed to
+    # avoid allocating unnecessary memory on the device.
+
+    # Note: we cannot combine the astype and to_device operations in one go
+    # using xp.asarray(..., dtype=dtype, device=device) because downcasting
+    # from float64 to float32 in asarray might not always be accepted as only
+    # casts following type promotion rules are guarateed to work.
+    # https://github.com/data-apis/array-api/issues/647
+    Q = to_device(Q, getattr(A, "device", None))
```

Comment:
I am concerned with `A.device` existing and `get_config()["array_api_dispatch"]` is False. If `array_api_compat` is not installed,  then `to_device` can fail.

As a workaround, I think we need to condition on `is_array_api_compliant` here or in `to_device`.

Reply:
Should be addressed. I re-ran the array_api test on a cuda machine to make sure that this still works as expected with GPU-backed libraries.